### PR TITLE
Merge carousel into master

### PR DESCRIPTION
**Pull request by mrdrummy93**

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+        this.state = {
+            position: 0,
+            page: 0,
+            animated: false,
+            isDown: false,
+            draggable: true,
+        }
+
+        this.colors = [];
```
Comment (KirylVasilyeuLMI) [inline src/App.js:20]: remove this

File: package.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "carousel-component",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
```
Comment (KirylVasilyeuLMI) [inline package.json:5]: index.js is not your entry file

File: package.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    "eslint-config-airbnb": "^18.2.1",
+    "eslint-plugin-babel": "^5.3.1",
+    "eslint-plugin-import": "^2.22.1",
+    "eslint-plugin-jsx-a11y": "^6.4.1",
+    "eslint-plugin-react": "^7.22.0",
+    "eslint-plugin-react-hooks": "^4.2.0",
+    "style-loader": "^0.23.1",
+    "webpack": "^5.22.0",
+    "webpack-cli": "^4.5.0",
+    "webpack-dev-server": "^3.11.2"
```
Comment (KirylVasilyeuLMI) [inline package.json:33]: remove ^ symbol on all dependencies

File: package.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  },
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
```
Comment (KirylVasilyeuLMI) [inline package.json:36]: are you using this script?

File: src/App.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
+import React from 'react';
+import {hot} from 'react-hot-loader';
+import Pagination from './Components/Pagination';
+
+
+class App extends React.Component {
+    constructor(props) {
+        super(props);
+
+        fool()
```
Comment (KirylVasilyeuLMI) [inline src/App.js:10]: delete this

File: src/App.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        this.colors = [];
+        for (let i = 0; i < 10; i++) {
+            this.colors.push(`rgb(${
+                Math.floor(Math.random() * 256)
+            }, ${
+                Math.floor(Math.random() * 256)
+            }, ${
+                Math.floor(Math.random() * 256)
+            })`)
+        }
```
Comment (KirylVasilyeuLMI) [inline src/App.js:29]: move all logic of children outside of this component

File: src/App.js [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        this.timeout = null;
+    }
+
+
+    renderArr = () => {
```
Comment (KirylVasilyeuLMI) [inline src/App.js:34]: bad naming

File: src/App.js [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        const arr = []
+        const {width = 400} = this.props
```
Comment (KirylVasilyeuLMI) [inline src/App.js:36]: don't use the default value

File: src/App.js [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+                    alignItems: 'center',
+                    display: "flex",
+                }} key={i}>
+                    <div style={{width: 200, height: 400, backgroundColor: this.colors[i]}}>
+                        <div>{i}</div>
+                    </div>
+                </div>
+            )
+        }
+        return arr;
```
Comment (KirylVasilyeuLMI) [inline src/App.js:51]: children's logic again

File: src/App.js [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    renderChild = () => {
+        const {width = 200} = this.props;
```
Comment (KirylVasilyeuLMI) [inline src/App.js:55]: default value

File: src/App.js [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+        const {page, position, animated} = this.state;
+        return (
+            <div style={{
+                display: "flex",
+                flexDirection: 'row',
+                transform: `translate(${-position}px, ${0}px)`,
+                marginLeft: page * width,
+                transition: animated ? 'all 300ms 0ms' : '',
```
Comment (KirylVasilyeuLMI) [inline src/App.js:63]: move all styles to styled components

File: src/App.js [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+        )
+    }
+
+    pointerStart = (xPosition) => {
+        if (this.state.draggable) {
+            this.setState({startPointerPosition: xPosition, animated: false})
+        }
+    }
+    pointerMove = (xPosition) => {
+        const {width = 200} = this.props;
```
Comment (KirylVasilyeuLMI) [inline src/App.js:76]: default value

File: src/App.js [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+        const position = this.state.startPointerPosition - xPosition
+
+        if (
+            this.state.startPointerPosition
+            && this.state.draggable
```
Comment (KirylVasilyeuLMI) [inline src/App.js:81]: a lot of destructuring can be used in whole file

File: src/App.js [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+            this.setState({position});
+        }
+    }
+
+    pointerEnd = () => {
+        if (this.state.startPointerPosition && this.state.draggable) {
+            this.setState({startPointerPosition: null, animated: true});
+            let page = this.state.page;
+
+            if (this.state.position < -40 && page < 0) {
```
Comment (KirylVasilyeuLMI) [inline src/App.js:99]: 40 should be some constants value or even should be customizable

File: src/App.js [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+                this.setState({position: -200, draggable: false});
```
Comment (KirylVasilyeuLMI) [inline src/App.js:100]: -width, i think

File: src/App.js [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    onTouchMove = (e) => this.pointerMove(e.touches[0].pageX)
+
+    onMouseMove = (e) => {
+        e.preventDefault();
+        this.pointerMove(e.pageX);
+    }
+
+    onClickLeft = () =>
+    {this.setState((prevState) => ({
+        page: Math.abs(prevState.page + 1) > this.colors.length - 10 ? prevState.page : prevState.page + 1,
```
Comment (KirylVasilyeuLMI) [inline src/App.js:128]: are you going to fix negative page values?

File: src/App.js [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+        animated: true
+    }))}
+
+    onClickRight = () =>
+    {this.setState((prevState) => ({
```
Comment (KirylVasilyeuLMI) [inline src/App.js:133]: you can write it like `foo = () => this.setState({})`, without additional {}

File: src/App.js [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+        page: Math.abs(prevState.page - 1) >= this.colors.length ? prevState.page : prevState.page - 1,
```
Comment (KirylVasilyeuLMI) [inline src/App.js:134]: you shouldn't lean on colors, it should be children's

File: src/App.js [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+    }))}
+
+    onClick = (i) => () => {
+        this.setState({page: -i, animated: true})}
+
+
+
+
+    render() {
+        const {width = 200} = this.props;
```
Comment (KirylVasilyeuLMI) [inline src/App.js:145]: default value

File: src/App.js [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+
+        const dots = [];
+        for (let i = 1; i <= this.colors.length; i++) {
```
Comment (KirylVasilyeuLMI) [inline src/App.js:148]: should be childrens

File: src/App.js [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+        return (
+            <>
+                <div className={'wrapper'} style={{
+                    flexDirection: 'row',
+                    display: 'inline-flex',
+                    margin: '0px auto',
+                    border: '1px solid black',
+                    offset: '10px',
+                    width: '100%',
+                    justifyContent: 'space-evenly',
```
Comment (KirylVasilyeuLMI) [inline src/App.js:161]: styles to styled components

File: src/App.js [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+                        onMouseMove={this.onMouseMove}
+                    >
+                        <div style={{
+                            margin: '40px',
+                            flexShrink: '0',
+                            outline: '5px solid black',
+                            flex: '1',
+                            justifyContent: 'center',
+                            overflow: 'hidden',
+                            width
```
Comment (KirylVasilyeuLMI) [inline src/App.js:179]: styles

File: src/App.js [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+                        }}>
+                            {this.renderChild()}
+                        </div>
+                    </div>
+                    <div style={{
+                        display: 'flex',
+                        flexDirection: 'row',
+                        position: 'absolute',
+                        top: '460px',
```
Comment (KirylVasilyeuLMI) [inline src/App.js:188]: styles

File: src/App.js [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+                        ))}
+                    </div>
+                    <button onClick={this.onClickRight}>❯</button>
+                </div>
+            </>
+        )
+    }
+}
+
+export default hot(module)(App);
```
Comment (KirylVasilyeuLMI) [inline src/App.js:201]: hot? why are you need this?

File: src/Components/Pagination.jsx [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+        return (
+            <button
+                style={{
+                    height: 15,
+                    width: 10,
+                    borderRadius: '50%',
+                    boxSizing: 'border-box',
+                    outline: 'none',
+                    marginRight: '5px',
+                }}
```
Comment (KirylVasilyeuLMI) [inline src/Components/Pagination.jsx:21]: styles

**Review by KirylVasilyeuLMI — COMMENTED** [review]
Your main target - move children outside of the component. and I think you should make `App.js` as main js file (instead `main.js`), and move carousel logic to some kind of `src/Components/Carousel/Carousel.js`, or near, `src/Carousel.js`. also you need an example folder, when you should have configured project

File: src/App.js (hunk 24 above)
Comment (mrdrummy93) [reply src/App.js:201]: Thanks to this app when i run my app changes to the code should update the client immediately after saving.

File: README.md [hunk 26]
```diff
@@ -1 +1,65 @@
...
+
+### Note
+
+`carousel-component` is using [styled-components](https://github.com/styled-components/styled-components) for styling. Don't forget to install it:
+
+```bash
+npm install --save styled-components
+```
+
+##Usage
```
Comment (KirylVasilyeuLMI) [inline README.md:24]: space

File: README.md [hunk 27]
```diff
@@ -1 +1,65 @@
...
+
+  return (
+    <Carousel width={ELEMENT_WIDTH} swipe={ELEMENT_SWIPE}>
+      {exampleItems.map(item => <div key={item.id}>{item.title}</div>)}
+    </Carousel>
+  )
+}
+```
+
+##Development
```
Comment (KirylVasilyeuLMI) [inline README.md:48]: space

File: README.md [hunk 28]
```diff
@@ -1 +1,65 @@
...
+
+```console
+git clone https://github.com/mrdrummy93/carousel-component.git
+cd carousel-component
+npm
+```
+
+###To run example
```
Comment (KirylVasilyeuLMI) [inline README.md:56]: space

File: README.md [hunk 29]
```diff
@@ -1 +1,65 @@
...
+```console
+npm example
+```
+
+The application is running at http://localhost:3000
+
+##License
```
Comment (KirylVasilyeuLMI) [inline README.md:63]: remove License

File: README.md [hunk 30]
```diff
@@ -1 +1,65 @@
...
+
+```angular2html
+npm install --save carousel-component
+```
+
+or
+
+```angular2html
+yarn add carousel-component
+```
```
Comment (KirylVasilyeuLMI) [inline README.md:14]: installation should be like development

File: README.md [hunk 31]
```diff
@@ -1 +1,65 @@
...
+  )
+}
+```
+
+##Development
+
+```console
+git clone https://github.com/mrdrummy93/carousel-component.git
+cd carousel-component
+npm
```
Comment (KirylVasilyeuLMI) [inline README.md:53]: npm install || yarn install

**Review by KirylVasilyeuLMI — COMMENTED** [review]
You should add a description for props as table

File: example/ExampleElement.jsx [hunk 32]
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+import Childrens from './Childrens';
+
+export default (props) => (
```
Comment (KirylVasilyeuLMI) [inline example/ExampleElement.jsx:4]: `export default ({ color }) => {`
and then you can write color={color} in render and remove eslint disabled rule

File: example/constants.js [hunk 33]
```diff
@@ -0,0 +1,3 @@
+export const ELEMENT_WIDTH = 200 + 10 + 10;
```
Comment (KirylVasilyeuLMI) [inline example/constants.js:1]: if you want to use margins here, create a new constant `MARGIN = 10`, and then `ELEMENT_WIDTH=200 + MARGIN * 2`

File: example/constants.js [hunk 34]
```diff
@@ -0,0 +1,3 @@
...
+export const ELEMENTS_COUNT = 10;
+export const ELEMENT_SWIPE = ELEMENT_WIDTH / 6;
```
Comment (KirylVasilyeuLMI) [inline example/constants.js:3]: `REQUIRED_SWIPE` or smthng like that

File: package.json (hunk 2 above)
Comment (KirylVasilyeuLMI) [reply package.json:5]: as I can see it still isn't you're entry file

File: package.json [hunk 35]
```diff
@@ -0,0 +1,50 @@
...
+    "webpack-dev-server": "3.11.2"
+  },
+  "scripts": {
+    "example": "webpack serve --mode development"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/mrdrummy93/carousel-component.git"
+  },
+  "author": "",
```
Comment (KirylVasilyeuLMI) [inline package.json:44]: put your nickname :)

File: src/Carousel.jsx [hunk 36]
```diff
@@ -0,0 +1,174 @@
...
+        }
+        this.setState({ position });
+      }
+    }
+
+    pointerEnd = () => {
+      const {
+        startPointerPosition, draggable, position,
+      } = this.state;
+      const { width, swipe } = this.props;
```
Comment (KirylVasilyeuLMI) [inline src/Carousel.jsx:74]: maybe you should add some kind of default value

Comment (KirylVasilyeuLMI) [reply src/Carousel.jsx:74]: like `const swipeToChangePage = swipe || width / 6`

File: src/Carousel.jsx [hunk 37]
```diff
@@ -0,0 +1,174 @@
...
+      }
+
+      return (
+        <CarouselWrapper
+          role="button"
+          tabIndex={0}
+          onTouchEnd={this.pointerEnd}
+          onMouseUp={this.pointerEnd}
+        >
+          <button type="button" onClick={this.onClickLeft}>❮</button>
```
Comment (KirylVasilyeuLMI) [inline src/Carousel.jsx:151]: some styles for buttons would be good

**Review by KirylVasilyeuLMI — COMMENTED** [review]
Are you using `npx create-react-app`? As I remember it's forbidden

**Comment by mrdrummy93** [comment]
> Are you using `npx create-react-app`? As I remember it's forbidden

Why do you think so

**Pull request merged**